### DISCUSSION TITLE
Async Transformer for Meilisearch

### DISCUSSION THREAD

**Discussion by RegisWeduLabs** [medusajs/medusa#4274] [Feature Requests]
Currently, using an async function to transform products in the medusa-plugin-meilisearch doesn't add the product to the indexes as the promises are returned instead of being awaited (without creating a custom meillisearch plugin).

I think it makes sense to add this feature, especially when talking about i18n where we might want to get some information on the product from an api - therefore async method needs to be used.  I will create the PR for the feature if this is something you want to consider?

**Comment by olivermrbl** [comment]
I think this is an excellent addition. Would love to accept a PR :)

**Comment by tsvetann** [comment]
@RegisHubelia this is an excellent initiative. @olivermrbl would it be possible to give more control to the developers when querying for products here such as passing a region and currency? This will allow is to index products with the calculated pricing for each variant along with price lists? I see that many devs are heavily investing in meilisearch and convinced this will be added value to their web apps.

**Comment by RegisWeduLabs** [comment]
@tsvetann, I'm not sure I understand what you mean...  I think having access to the medusa "container" that contains all the service classes etc should give the users enough leverage to do what they want?  One thing tough, I just realised that the code actually checks if the index is the products index - and then send the documents to the transformer:

```
  async getTransformedDocuments(type: string, documents: any[]) {
    if (!documents?.length) {
      return []
    }
    switch (type) {
      case SearchTypes.indexTypes.PRODUCTS:
        const productsTransformer =
          this.config_.settings?.[SearchTypes.indexTypes.PRODUCTS]
            ?.transformer ?? transformProduct
        const results = await Promise.all(documents.map(productsTransformer))

        return await Promise.all(documents.map(productsTransformer))
      default:
        return documents
    }
  }
```

Is there a reason for this?  Because basically - that would mean that transformers can only be run on the products indexes and not any other indexes. 

From the Docs:
![screenshot_2023-06-12-120311](https://github.com/medusajs/medusa/assets/25985730/2e462ac4-4749-4211-98cf-1f252754ed5d)

But this is inacurate, again, to my understanding.  I think there are few quick features that can be added to give much more flexibility to the plugin - my ideas would be : support async (I've done that), provide the medusa container to the transformer function so that it can be used for different use cases, and allow tranformer for any other indexes, including custom indexes that could fetch or get data from different sources.

I would therefore propose to change the current function:

```
  async getTransformedDocuments(type: string, documents: any[]) {
    if (!documents?.length) {
      return []
    }

    switch (type) {
      case SearchTypes.indexTypes.PRODUCTS:
        const productsTransformer =
          this.config_.settings?.[SearchTypes.indexTypes.PRODUCTS]
            ?.transformer ?? transformProduct

        return await Promise.all(documents.map(productsTransformer))
      default:
        return documents
    }
  }
```

to this (of course adding awaits where the function is called in the service file:

```
  async getTransformedDocuments(type: string, documents: any[]) {
    if (!documents?.length) {
      return []
    }

    const transformer =
      this.config_.settings?.[type]?.transformer ??
      (type === SearchTypes.indexTypes.PRODUCTS ? transformProduct : (document) => document)

    const results = await Promise.allSettled(documents.map(transformer));
    const rejected = <T,>(p: PromiseSettledResult<T>): p is PromiseRejectedResult => p.status === 'rejected';

    const errors = results.filter(rejected);
    if (errors.length) {
        console.error('An error occurred while transforming some documents:', errors, `for index ${type}`);
    }

    const fulfilled = <T,>(p:PromiseSettledResult<T>): p is PromiseFulfilledResult<T> => p.status === 'fulfilled';
    return results.filter(fulfilled).map((r) => r.value);
  }
```

This way, you can actually add your own indexes, async, and I think this can help with the i18n not being supported...  Calling a CMS to get other languages can add significant delays on the front end side - whereas if you use the meilisearch indexes for multiple languages, you still get the speed and flexibility...

Now this doesn't include the medusa container...  The issue I have here is that we'd need to send the container as a first argument, so it would break current implementations - or we'd need to pass it as a second argument, but this isn't standard looking at other plugins... 

Toughts?

**Reply by tsvetann** [reply]
Yes basically agree with everything you said. I made a preliminary PR to discuss this approach with the Medusa team

**Reply by RegisWeduLabs** [reply]
Okay - so I won't bother creating a PR then?

**Reply by RegisWeduLabs** [reply]
Also - just checked the PR - but I think some kind of error handling like the example above using allSettled instead of .all would make error handling clearer - especially if using external sources that could cause errors - it would be useful to know if there is an issue...  Plus - it wouldn't work with all indexes (customs especially).

**Reply by tsvetann** [reply]
Thanks for the feedback. Can you please add your comments on the PR so we can consolidate the discussion? Thanks
